### PR TITLE
more pattern matching

### PR DESCRIPTION
**Pull request by STRd6**
I removed the separate `Match...` rules and expanded `BindingPattern` this way it will be the same parsing for decs and patterns. I was also able to simplify some of the pattern matching cases. And finally I added pinning. There's still another batch of work that needs to be done to finish https://github.com/DanielXMoore/Civet/issues/335 but the foundational pieces are in place with this PR.

### CODE REVIEW

File: test/switch.civet [hunk 1]
```diff
@@ -612,7 +648,7 @@ describe "switch", ->
           console.log(b)}
     """
 
-    testCase.skip """
+    testCase """
```
Comment (edemaine) [inline test/switch.civet:651]: Maybe add a test with other binary operators, along the lines of #335?
```coffee
switch x
  < 5
```

**Review by edemaine — APPROVED** [review]
So cool to see `Match` and `Binding` rules unified!

It looks like `{x, ^y} := obj` currently compiles to `const {x, y} = obj`. We should at least add a to-do or a skipped test about that... Or perhaps even an exception?

`if {x, ^y} := obj` is even more confusing at this point. But that's likely a new issue to raise. (We're planning for pattern matching in that case, right?)

File: source/parser.hera [hunk 2]
```diff
@@ -1110,28 +1117,32 @@ AtIdentifierRef
...
 ObjectBindingPattern
-  OpenBrace ObjectBindingPatternContent:c __ CloseBrace ->
+  _? OpenBrace ObjectBindingPatternContent:c __ CloseBrace ->
     return {
       type: "ObjectBindingPattern",
-      children: [$1, ...c.children, $3, $4],
+      children: $0,
       content: c.children,
       names: c.names,
+      properties: c.children,
```
[the remark this answers is not in the document]
Comment (STRd6) [reply source/parser.hera:1145]: We don't need both, I just haven't consolidated them yet, thanks for the reminder!

File: source/parser.hera [hunk 3]
```diff
@@ -1218,31 +1231,49 @@ BindingProperty
...
       init,
-      names: b.names,
-      children: $0
+      names: value.names,
     }
 
-  BindingIdentifier:b Initializer?:init ->
-    if (b.type === "AtBinding") {
+  _?:ws "^"?:pin BindingIdentifier:binding Initializer?:init ->
+    // TODO make this work with pin
```
[the remark this answers is not in the document]
Comment (STRd6) [reply source/parser.hera:1244]: Unifying the pattern matching with object literal syntax sounds like a good idea. I'll add it to the list in the pattern matching issue and do it as a follow up.

File: source/parser.hera [hunk 4]
```diff
@@ -1110,28 +1117,32 @@ AtIdentifierRef
       id: id.name,
     }
 
-# https://262.ecma-international.org/#prod-BindingPattern
-BindingPattern
-  __ ObjectBindingPattern:p ->
+PinPattern
+  "^" Identifier:identifier ->
```
[the remark this answers is not in the document]
Comment (STRd6) [reply source/parser.hera:1121]: That seems like a good idea, I'll do it in a follow up.

**Comment by STRd6** [comment]
> It looks like `{x, ^y} := obj` currently compiles to `const {x, y} = obj`. We should at least add a to-do or a skipped test about that... Or perhaps even an exception?
> 
> `if {x, ^y} := obj` is even more confusing at this point. But that's likely a new issue to raise. (We're planning for pattern matching in that case, right?)

I'll add these to the follow up todo list as well.

**Pull request merged**